### PR TITLE
SDN-4919: NetworkSegmentation: Add tests for EndpointSlices mirror controller

### PR DESCRIPTION
**Pull request by kyrtapz**
Based on https://github.com/openshift/origin/pull/28964.
Depends on https://github.com/openshift/ovn-kubernetes/pull/2241

This PR refactors and adds tests for NetworkSegmentation featuregate which is currently in TechPreview.

### CODE REVIEW

**Comment by tssurya** [comment]
/retitle SDN-4919: NetworkSegmentation: Add tests for EndpointSlices mirror controller

File: test/extended/networking/network_segmentation_endpointslice_mirror.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eservice "k8s.io/kubernetes/test/e2e/framework/service"
+	admissionapi "k8s.io/pod-security-admission/api"
+	"k8s.io/utils/net"
```
Comment (maiqueb) [inline test/extended/networking/network_segmentation_endpointslice_mirror.go:20]: nit: consider splitting these . Not a must.

File: test/extended/networking/network_segmentation_endpointslice_mirror.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+					Expect(nodeConfig).To(HaveLen(1))
+					v4HostSubnet, v6HostSubnet = nodeConfig[0].IFAddr.IPv4, nodeConfig[0].IFAddr.IPv6
+				}
+
+				isDualStack := v4HostSubnet != "" && v6HostSubnet != ""
+
+				By("asserting the mirrored EndpointSlice exists and contains PODs primary IPs")
+				Eventually(func() error {
+					return validateMirroredEndpointSlices(cs, f.Namespace.Name, svc.Name, userDefinedNetworkIPv4Subnet, userDefinedNetworkIPv6Subnet, v4HostSubnet, v6HostSubnet, replicas, isDualStack, isHostNetwork)
+
```
Comment (maiqueb) [inline test/extended/networking/network_segmentation_endpointslice_mirror.go:108]: nit: this extra line is confusing. either drop it, or put all `validateMirroredEndpointSlices` args in a separate line

File: test/extended/networking/network_segmentation_endpointslice_mirror.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+					esList, err := cs.DiscoveryV1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", "k8s.ovn.org/service-name", svc.Name)})
+					if err != nil {
+						return err
+					}
+
+					if len(esList.Items) != 0 {
+						return fmt.Errorf("expected no mirrored EndpointSlice, got: %d", len(esList.Items))
+					}
+					return nil
+				}, 2*time.Minute, 6*time.Second).ShouldNot(HaveOccurred())
```
Comment (maiqueb) [inline test/extended/networking/network_segmentation_endpointslice_mirror.go:231]: nit: positive logic reads better IMHO.

```suggestion
				}, 2*time.Minute, 6*time.Second).Should(Succeed())
```

**Review by maiqueb — APPROVED** [review]
just some nits.

**Comment by kyrtapz** [comment]
/hold
we need to wait for the next ovn-kubernetes downstream merge

**Comment by kyrtapz** [comment]
/retest-required
/test e2e-gcp-ovn-techpreview
/test e2e-aws-ovn-single-node-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.17-e2e-aws-ovn-techpreview

**Comment by tssurya** [comment]
tech preview has passed: https://pr-payload-tests.ci.openshift.org/runs/ci/7fd1def0-50e8-11ef-95d9-ea538a99f7fb-0

**Comment by tssurya** [comment]
gcp has also passed: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/28945/pull-ci-openshift-origin-master-e2e-gcp-ovn-techpreview/1819410719368548352

**Comment by neisw** [comment]
Just curious, are you expecting the tests to flake?  I see the passes in gcp but aws has flakes/failures  https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview/1819403219323850752

```
: [sig-network][OCPFeatureGate:NetworkSegmentation][Feature:UserDefinedPrimaryNetworks] EndpointSlices mirroring when using openshift ovn-kubernetes mirrors EndpointSlices managed by the default controller for namespaces with user defined primary networks L2 dualstack primary UDN, host-networked pods [Suite:openshift/conformance/parallel] expand_less
Run #0: Failed expand_less 	2m41s
{  fail [github.com/openshift/origin/test/extended/networking/network_segmentation_endpointslice_mirror.go:110]: Timed out after 120.000s.
Expected success, but got an error:
    <*errors.errorString | 0xc0017dcba0>: 
    ip [10.0.64.93] is NOT in range 10.0.0.0/18
    {
        s: "ip [10.0.64.93] is NOT in range 10.0.0.0/18",
    }
Ginkgo exit error 1: exit with code 1}
[open stdoutopen_in_new](https://prow.ci.openshift.org/spyglass/lens/junit/iframe?req=%7B%22artifacts%22%3A%5B%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_install_status.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_symptoms.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-must-gather%2Fartifacts%2Fjunit_install.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fe2e-monitor-tests__20240802-165800.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fjunit_e2e__20240802-165800.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_node_ready.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_nodes.xml%22%2C%22artifacts%2Fjunit_operator.xml%22%5D%2C%22index%22%3A2%2C%22src%22%3A%22gs%2Ftest-platform-results%2Flogs%2Fopenshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview%2F1819403219323850752%22%7D&topURL=https%3A//prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview/1819403219323850752&lensIndex=2#)
Run #1: Passed expand_less 	33s
<nil>
```

```
: [sig-network][OCPFeatureGate:NetworkSegmentation][Feature:UserDefinedPrimaryNetworks] EndpointSlices mirroring when using openshift ovn-kubernetes mirrors EndpointSlices managed by the default controller for namespaces with user defined primary networks L3 dualstack primary UDN, host-networked pods [Suite:openshift/conformance/parallel] expand_less
Run #0: Failed expand_less 	2m22s
{  fail [github.com/openshift/origin/test/extended/networking/network_segmentation_endpointslice_mirror.go:110]: Timed out after 120.000s.
Expected success, but got an error:
    <*errors.errorString | 0xc001ea6970>: 
    ip [10.0.64.93] is NOT in range 10.0.0.0/18
    {
        s: "ip [10.0.64.93] is NOT in range 10.0.0.0/18",
    }
Ginkgo exit error 1: exit with code 1}
[open stdoutopen_in_new](https://prow.ci.openshift.org/spyglass/lens/junit/iframe?req=%7B%22artifacts%22%3A%5B%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_install_status.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_symptoms.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-must-gather%2Fartifacts%2Fjunit_install.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fe2e-monitor-tests__20240802-165800.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fjunit_e2e__20240802-165800.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_node_ready.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_nodes.xml%22%2C%22artifacts%2Fjunit_operator.xml%22%5D%2C%22index%22%3A2%2C%22src%22%3A%22gs%2Ftest-platform-results%2Flogs%2Fopenshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview%2F1819403219323850752%22%7D&topURL=https%3A//prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview/1819403219323850752&lensIndex=2#)
Run #1: Passed expand_less 	33s
<nil>
```

**Comment by kyrtapz** [comment]
> Just curious, are you expecting the tests to flake? I see the passes in gcp but aws has flakes/failures https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview/1819403219323850752
> 
> ```
> : [sig-network][OCPFeatureGate:NetworkSegmentation][Feature:UserDefinedPrimaryNetworks] EndpointSlices mirroring when using openshift ovn-kubernetes mirrors EndpointSlices managed by the default controller for namespaces with user defined primary networks L2 dualstack primary UDN, host-networked pods [Suite:openshift/conformance/parallel] expand_less
> Run #0: Failed expand_less 	2m41s
> {  fail [github.com/openshift/origin/test/extended/networking/network_segmentation_endpointslice_mirror.go:110]: Timed out after 120.000s.
> Expected success, but got an error:
>     <*errors.errorString | 0xc0017dcba0>: 
>     ip [10.0.64.93] is NOT in range 10.0.0.0/18
>     {
>         s: "ip [10.0.64.93] is NOT in range 10.0.0.0/18",
>     }
> Ginkgo exit error 1: exit with code 1}
> [open stdoutopen_in_new](https://prow.ci.openshift.org/spyglass/lens/junit/iframe?req=%7B%22artifacts%22%3A%5B%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_install_status.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_symptoms.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-must-gather%2Fartifacts%2Fjunit_install.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fe2e-monitor-tests__20240802-165800.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fjunit_e2e__20240802-165800.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_node_ready.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_nodes.xml%22%2C%22artifacts%2Fjunit_operator.xml%22%5D%2C%22index%22%3A2%2C%22src%22%3A%22gs%2Ftest-platform-results%2Flogs%2Fopenshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview%2F1819403219323850752%22%7D&topURL=https%3A//prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview/1819403219323850752&lensIndex=2#)
> Run #1: Passed expand_less 	33s
> <nil>
> ```
> 
> ```
> : [sig-network][OCPFeatureGate:NetworkSegmentation][Feature:UserDefinedPrimaryNetworks] EndpointSlices mirroring when using openshift ovn-kubernetes mirrors EndpointSlices managed by the default controller for namespaces with user defined primary networks L3 dualstack primary UDN, host-networked pods [Suite:openshift/conformance/parallel] expand_less
> Run #0: Failed expand_less 	2m22s
> {  fail [github.com/openshift/origin/test/extended/networking/network_segmentation_endpointslice_mirror.go:110]: Timed out after 120.000s.
> Expected success, but got an error:
>     <*errors.errorString | 0xc001ea6970>: 
>     ip [10.0.64.93] is NOT in range 10.0.0.0/18
>     {
>         s: "ip [10.0.64.93] is NOT in range 10.0.0.0/18",
>     }
> Ginkgo exit error 1: exit with code 1}
> [open stdoutopen_in_new](https://prow.ci.openshift.org/spyglass/lens/junit/iframe?req=%7B%22artifacts%22%3A%5B%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_install_status.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_symptoms.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fgather-must-gather%2Fartifacts%2Fjunit_install.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fe2e-monitor-tests__20240802-165800.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fjunit_e2e__20240802-165800.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_node_ready.xml%22%2C%22artifacts%2Fe2e-aws-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_nodes.xml%22%2C%22artifacts%2Fjunit_operator.xml%22%5D%2C%22index%22%3A2%2C%22src%22%3A%22gs%2Ftest-platform-results%2Flogs%2Fopenshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview%2F1819403219323850752%22%7D&topURL=https%3A//prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview/1819403219323850752&lensIndex=2#)
> Run #1: Passed expand_less 	33s
> <nil>
> ```

Good catch! this was not expected and I've fixed the issue(AWS uses different EIP subnets per node).
Additionally modified the east-west tests to work on SNO.

**Comment by kyrtapz** [comment]
/test e2e-gcp-ovn-techpreview
/test e2e-aws-ovn-single-node-techpreview
/payload-job periodic-ci-openshift-release-master-ci-4.17-e2e-aws-ovn-techpreview

File: test/extended/networking/network_segmentation.go [hunk 4]
```diff
@@ -79,13 +76,12 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
...
 				Expect(err).NotTo(HaveOccurred())
-				// TODO: should I skip this on SNO ?... or should I simply put pods the first and last node ?
-				Expect(len(workerNodes)).To(BeNumerically(">=", 2))
+				Expect(len(workerNodes)).To(BeNumerically(">=", 1))
 
 				clientPodConfig.namespace = f.Namespace.Name
 				clientPodConfig.nodeSelector = map[string]string{nodeHostnameKey: workerNodes[0].Name}
 				serverPodConfig.namespace = f.Namespace.Name
-				serverPodConfig.nodeSelector = map[string]string{nodeHostnameKey: workerNodes[1].Name}
+				serverPodConfig.nodeSelector = map[string]string{nodeHostnameKey: workerNodes[len(workerNodes)-1].Name}
```
Comment (maiqueb) [inline test/extended/networking/network_segmentation.go:84]: OK, if you're doing it here, I'll just close #28966. Thanks.

**Comment by kyrtapz** [comment]
All techpreview jobs are green, pushing again to address the last AWS flake:
```
: [sig-network][OCPFeatureGate:NetworkSegmentation][Feature:UserDefinedPrimaryNetworks] when using openshift ovn-kubernetes is isolated from the default network with L3 dualstack primary UDN [Suite:openshift/conformance/parallel] expand_less
Run #0: Failed expand_less 	34s
{  fail [github.com/openshift/origin/test/extended/networking/network_segmentation.go:278]: Expected
    <bool>: false
to be true
Ginkgo exit error 1: exit with code 1}
```

**Comment by kyrtapz** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.17-e2e-aws-ovn-techpreview

**Comment by kyrtapz** [comment]
techpreview aws is green with no flakiness: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-origin-28945-ci-4.17-e2e-aws-ovn-techpreview/1820437926434050048

**Comment by kyrtapz** [comment]
/test e2e-gcp-ovn-techpreview
/test e2e-aws-ovn-single-node-techpreview
/payload-job periodic-ci-openshift-release-master-ci-4.17-e2e-aws-ovn-techpreview

**Comment by kyrtapz** [comment]
`e2e-aws-ovn-edge-zones` seems to be [failing today](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-openshift-origin-master-e2e-aws-ovn-edge-zones) with:
`node count should match or exceed machine count` 
This happens across all PRs and doesn't seem related to the changes introduced here.

**Comment by neisw** [comment]
/override ci/prow/e2e-aws-ovn-edge-zones

previously passed on the same commit

**Comment by neisw** [comment]
/override ci/prow/e2e-aws-ovn-serial
/override ci/prow/e2e-gcp-ovn

previously passed on the same commit

**Pull request merged**